group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java

```diff
@@ -0,0 +1,500 @@
+
+        ConsumerGroupMember member2 = new ConsumerGroupMember.Builder("member2")
+            .setMemberEpoch(1)
+            .setPreviousMemberEpoch(0)
+            .setNextMemberEpoch(1)
+            .build();
+
+        consumerGroup.updateMember(member2);
+        consumerGroup.setGroupEpoch(2);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.ASSIGNING, consumerGroup.state());
+
+        consumerGroup.setAssignmentEpoch(2);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.RECONCILING, consumerGroup.state());
+
+        member1 = new ConsumerGroupMember.Builder(member1)
+            .setMemberEpoch(2)
+            .setPreviousMemberEpoch(1)
+            .setNextMemberEpoch(2)
+            .build();
+        consumerGroup.updateMember(member1);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.RECONCILING, consumerGroup.state());
+
+        member2 = new ConsumerGroupMember.Builder(member2)
+            .setMemberEpoch(2)
+            .setPreviousMemberEpoch(1)
+            .setNextMemberEpoch(2)
+            .build();
+        consumerGroup.updateMember(member2);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.STABLE, consumerGroup.state());
+
+        consumerGroup.removeMember("member1");
+        consumerGroup.removeMember("member2");
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.EMPTY, consumerGroup.state());
+    }
+
+    @Test
+    public void testPreferredServerAssignor() {
+        ConsumerGroup consumerGroup = createConsumerGroup("foo");
+
+        ConsumerGroupMember member1 = new ConsumerGroupMember.Builder("member1")
+            .setServerAssignorName("range")
+            .build();
+        ConsumerGroupMember member2 = new ConsumerGroupMember.Builder("member2")
+            .setServerAssignorName("range")
+            .build();
+        ConsumerGroupMember member3 = new ConsumerGroupMember.Builder("member3")
+            .setServerAssignorName("uniform")
+            .build();
+
+        assertEquals(
+            Optional.empty(),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, member1)
+        );
+
+        consumerGroup.updateMember(member1);
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        assertEquals(
+            Optional.empty(),
+            consumerGroup.preferredServerAssignor(member1, null)
+        );
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, member2)
+        );
+
+        consumerGroup.updateMember(member2);
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        consumerGroup.updateMember(member3);
```

Comment:
If we have more than one assignor with the same max value, we choose one randomly (based on map iteration order). Is this intended?

Reply:
Ah I see this below.

Reply:
Correct.